google/cloud/bigtable/internal/rate_limiter.cc

```diff
@@ -0,0 +1,80 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/bigtable/internal/rate_limiter.h"
+#include "absl/time/time.h"
+
+namespace google {
+namespace cloud {
+namespace bigtable_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using Clock = RateLimiter::Clock;
+
+/**
+ * Returns the duration it takes to give out @p permits at the given @p rate.
+ *
+ * Essentially, we are returning `permits / rate`.
+ *
+ * This function handles rounding, while getting the maximum precision out of
+ * the clock.
+ */
+Clock::duration TimeNeeded(std::int64_t permits, double rate) {
+  return std::chrono::duration_cast<Clock::duration>(absl::ToChronoNanoseconds(
+      absl::Seconds(static_cast<double>(permits) / rate)));
+}
+
+}  // namespace
+
+RateLimiter::Clock::duration RateLimiter::acquire(std::int64_t permits) {
+  auto const now = clock_->Now();
+  std::lock_guard<std::mutex> lk(mu_);
```

Comment:
What is this protecting? Does this critical section need to be as long as it is?

Reply:
> What is this protecting?

It is protecting `next_`, `stored_permits_` , `period_` (formerly known as `rate_`).

> Does this critical section need to be as long as it is?

I need to grab the lock before using any of the values. And I need to hold the lock until I am done updating those values. This "critical section" is just doing arithmetic, so it should be snappy.

I could unlock the lock before returning the values, but I assume that is a negligible difference.

Reply:
You might want to consider whether you really need to lock all those values. 

I don't know enough about this change to give a suggestion though

Reply:
I really do need to lock them. This test is representative of how an application might use a client. If I remove the lock, the test starts failing. We end up letting too many requests through.

https://github.com/googleapis/google-cloud-cpp/blob/c61e53b42e45fb1ee8dfa4cb62ca8093ead4e35c/google/cloud/bigtable/internal/rate_limiter_test.cc#L177